### DISCUSSION TITLE
Native: OpenGL Optimization

### DISCUSSION THREAD

**Discussion by sjg-wdw** [maplibre/maplibre#331] [Native feature proposals]
## OpenGL Optimization for Android

Here's some functionality we'd like to add to MapLibre. This is one proposal of several for Native and if you'd like a little context and our strategy [there's an overview](https://www.wetdogweather.com/blog/maplibre-proposals). This one is purely Native.

When we upgraded MapLibre to a modular rendering architecture, we also moved the OpenGL support to ES 3.0.  That opens up a lot of possibilities for new functionality, of course, but also optimization.  You can simply do things more efficiently with 3.x than you could with 2.x.

This doesn't matter for iOS or MacOS now that we're moving to Metal.  But it does still matter for Android and there are significant improvements to be made.

### Motivation

Rendering speed, frames per second, is always the most obvious motivation, but it's not most important on mobile devices.  Power is the thing that matters and power usage tends to track frame time fairly well.  The short version is: use the GPU more, use the CPU less.  And just generally do less.

Better rendering performance also allows developers to do more on top of the toolkit.  If MapLibre isn't hogging the CPU, a developer can do their work more easily.

That's the 'why', now for the 'how'.

### Data Driven Styling & The GPU

Much of the thrashing on the current implementation come from visual changes based on zoom level and other inputs.  Whenever the user zooms in or out a variety of features are tweaking their opacity or color or width or a set of other options.  All this tweaking is happening on the CPU.  More of it should be happening on the CPU.

We're just starting this shift for Metal on iOS, a platform that has some pretty nice analysis tools for understanding what's going on.  There are two parts to this:
- Teasing out what the parameter update functions are
- Implementing those functions in the shaders

We won't be able to get all of those data driven styling options into the shaders. Some control how the geometry is generated or fundamentally change the look and feel, like pointing to a new font.  But most of the more incremental functions used in the course of map display should be possible.

Once we're done with the Metal version, it would be possible to take that knowledge and apply it to OpenGL ES 3 on Android.

### Instancing

The toolkit was originally written in the OpenGL ES 2.0 era with OpenGL ES 1.0 on their minds. Now that 3.0 is widely supported there are new options available.  Instancing is a big one.

As it stands, each symbol is represented by four points and two triangles when rendered.  There are a whole lot of symbols in some of the test areas and they could all be instanced.

Instancing results in one entry per symbol, a potentially huge savings in memory.  Doing more of the work on the GPU lets you merge disparate groups of symbols together, but at a cost of memory per instance.  It's a balancing act, but we could potentially see both memory savings and increased performance.

### Analysis & Optimization

The OpenGL version of MapLibre would benefit from more performance analysis, particularly on Android.  The biggest cost we're seeing currently is in shuffling buffer objects around to feed the uniforms that shaders need. Fewer of those buffers would be a big benefit and we can achieve that by doing more on the GPU.

Beyond that, the toolkit would benefit from a few more analysis/optimization cycles.

### Discussion

Beyond the basics of moving more work to the GPU, just focusing some skilled developer time on the Android renderer would have benefit.  We've seen a lot of opportunity for better resource usage as we restructure things for Metal.

**Comment by lseelenbinder** [comment]
What's the advantage/trade-offs between this proposal and the Vulkan proposal? If we opt to implement Vulkan, would that obviate this proposal (or are there enough platforms that would still use OpenGL to make it worth while)?

**Reply by sjg-wdw** [reply]
OpenGL optimization is going to be cheaper than doing Vulkan and I'm not sure what Vulkan adoption looks like on Android.

As for other platforms, Windows I believe.  Probably Linux too.  I know there's Vulkan there, but it may not be heavily used.

**Comment by thehoneymad** [comment]
Interesting, I am onboard with the idea of doing more in the GPU over CPU. But on the Vulkan vs OpenGL front, I still wish to keep the door open to Vulkan because if I am not wrong, [Android is keeping OpenGL ES in keep the lights on mode](https://developer.android.com/games/develop/use-vulkan). If we were to do this, we would want to carve a path towards Vulkan so the ideas of @sjg-wdw doesnt get jailed in the future.

There's much to learn from what has happaned here. I will stay tuned here more to learn.

**Reply by sjg-wdw** [reply]
There's a class of devices that wants to be primarily Vulkan, so that's something to consider.  But I think, realistically, OpenGL will be around for quite a while.

**Reply by thehoneymad** [reply]
I agree to that sentiment. As long as we have a path forward that we can act on for Vulkan, I am aligned with your proposal. 

Thinking lonb term, if time is what we lack to do Vulkan, we can discuss about that. :)
